File: scheduler/internal/tracing/tracing.go
// Package tracing OpenTelemetry 链路：三进程统一 trace_id（design 14）。
//
// 环境变量：
//
//	TP_OTEL_EXPORTER  "" 关闭（默认）| "stdout" 开发调试 | "otlp" 发 Collector
//	TP_OTEL_ENDPOINT  otlp gRPC 地址（默认 127.0.0.1:4317，dev 一律 insecure）
//
// 传播：Scheduler REST 入口由下方 fiber 中间件起 span、gRPC 由 otelgrpc 起 span；
// 派发给 Worker 时 traceparent 写入 TaskAssignment（Worker 侧提取续链）；
// Copilot 经 gRPC metadata 注入。关闭时 InjectTraceparent 返回 ""，全链路零开销。
package tracing

import (
	"context"
	"os"

	"github.com/gofiber/fiber/v3"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

func init() {
	// 传播器始终安装：即使 SDK 关闭，注入/提取逻辑也安全（返回空 carrier）。
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{}, propagation.Baggage{}))
}

// Init 按环境初始化 TracerProvider；返回 shutdown（main defer 调用）。关闭时为 no-op。
func Init(serviceName string) func(context.Context) {
	switch os.Getenv("TP_OTEL_EXPORTER") {
	case "stdout", "otlp":
	default:
		return func(context.Context) {}
	}

	// NewSchemaless：避免与 resource.Default() 的 semconv schema URL 冲突
	// （冲突会让 Merge 返回 error 并丢掉 service.name，回退成 unknown_service）。
	res, err := resource.Merge(resource.Default(),
		resource.NewSchemaless(semconv.ServiceName(serviceName)))
	if err != nil {
		res = resource.Default()
	}
	if err != nil {
		res = resource.Default()
	}

	var exp sdktrace.SpanExporter
	if os.Getenv("TP_OTEL_EXPORTER") == "stdout" {
		exp, err = stdouttrace.New()
	} else {
		endpoint := os.Getenv("TP_OTEL_ENDPOINT")
		if endpoint == "" {
			endpoint = "127.0.0.1:4317"
		}
		exp, err = otlptracegrpc.New(context.Background(),
			otlptracegrpc.WithEndpoint(endpoint), otlptracegrpc.WithInsecure())
	}
	if err != nil {
		return func(context.Context) {}
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithResource(res),
		sdktrace.WithBatcher(exp),
	)
	otel.SetTracerProvider(tp)
	return func(ctx context.Context) { _ = tp.Shutdown(ctx) }
}

// InjectTraceparent 提取 ctx 当前 span 的 W3C traceparent（无有效 span → ""）。
func InjectTraceparent(ctx context.Context) string {
	carrier := propagation.MapCarrier{}
	otel.GetTextMapPropagator().Inject(ctx, carrier)
	return carrier["traceparent"]
}

// TraceID 日志关联用（无有效 span → ""）。
func TraceID(ctx context.Context) string {
	sc := trace.SpanContextFromContext(ctx)
	if !sc.HasTraceID() {
		return ""
	}
	return sc.TraceID().String()
}

// Middleware 起 HTTP server span：提取入站 traceparent（REST 侧通常无父），
// span ctx 经 c.SetContext 供下游（auth/runner.Trigger 等）续链；Next 后回填
// 路由模板与状态码。SDK 关闭时 Start 返回 no-op span，零开销。
func Middleware() fiber.Handler {
	tracer := otel.Tracer("testpilot/scheduler")
	return func(c fiber.Ctx) error {
		ctx := otel.GetTextMapPropagator().Extract(c.Context(), headerCarrier{c})
		ctx, span := tracer.Start(ctx, "scheduler.http", trace.WithSpanKind(trace.SpanKindServer))
		c.SetContext(ctx)
		defer span.End()
		err := c.Next()
		route := "unmatched"
		if r := c.Route(); r != nil && r.Path != "" && r.Path != "/" {
			route = r.Path
		}
		status := c.Response().StatusCode()
		span.SetName(c.Method() + " " + route)
		span.SetAttributes(
			attribute.String("http.request.method", c.Method()),
			attribute.String("http.route", route),
			attribute.Int("http.response.status_code", status),
		)
		if status >= fiber.StatusInternalServerError {
			span.SetStatus(codes.Error, "")
		}
		return err
	}
}

// headerCarrier 仅供 Extract：propagator 按其关心的 key（traceparent/tracestate/baggage）
// 逐个 Get，无需实现 Set/Keys。
type headerCarrier struct{ c fiber.Ctx }

func (h headerCarrier) Get(key string) string { return h.c.Get(key) }
func (h headerCarrier) Set(string, string)    {}
func (h headerCarrier) Keys() []string        { return nil }

File: worker/src/testpilot_worker/tracing.py
"""OpenTelemetry 链路（Worker 侧）：与 Scheduler 统一 trace_id（design 14）。

环境变量：
  TP_OTEL_EXPORTER  "" 关闭（默认）| "stdout" 开发调试 | "otlp" 发 Collector
  TP_OTEL_ENDPOINT  otlp gRPC 地址（默认 127.0.0.1:4317）

Scheduler 派发任务时在 TaskAssignment.traceparent 携带 W3C traceparent，
本模块 extract() 提取续链；LogFilter 把当前 span 的 trace_id 注入日志记录，
无 span 时字段为空字符串（格式稳定，便于 grep/采集）。
"""

from __future__ import annotations

import logging
import os

from opentelemetry import trace
from opentelemetry.propagate import extract as _otel_extract
from opentelemetry.propagate import inject as _otel_inject
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter

tracer = trace.get_tracer("testpilot.worker")


def init(service_name: str = "testpilot-worker") -> None:
    """按 TP_OTEL_EXPORTER 初始化 TracerProvider；默认关闭（零开销）。"""
    mode = os.environ.get("TP_OTEL_EXPORTER", "")
    if mode not in ("stdout", "otlp"):
        return
    if mode == "stdout":
        exporter = ConsoleSpanExporter()
    else:
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

        exporter = OTLPSpanExporter(
            endpoint=os.environ.get("TP_OTEL_ENDPOINT", "127.0.0.1:4317"), insecure=True
        )
    provider = TracerProvider(resource=Resource.create({"service.name": service_name}))
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)


def extract_context(traceparent: str):
    """从 traceparent 字符串提取上游上下文（空 → 全新 trace）。"""
    if not traceparent:
        return None
    return _otel_extract({"traceparent": traceparent})


def current_traceparent() -> str:
    """当前 span 的 traceparent（无 → ""）。"""
    carrier: dict[str, str] = {}
    _otel_inject(carrier)
    return carrier.get("traceparent", "")


def current_trace_id() -> str:
    span = trace.get_current_span()
    sc = span.get_span_context()
    return format(sc.trace_id, "032x") if sc and sc.trace_id else ""


class TraceLogFilter(logging.Filter):
    """给日志记录注入 trace_id / span_id 字段。"""

    def filter(self, record: logging.LogRecord) -> bool:
        span = trace.get_current_span()
        sc = span.get_span_context() if span else None
        record.trace_id = format(sc.trace_id, "032x") if sc and sc.trace_id else ""
        record.span_id = format(sc.span_id, "016x") if sc and sc.span_id else ""
        return True


def attach_log_filter() -> None:
    """root logger 所有 handler 挂 TraceLogFilter（basicConfig 之后调用）。"""
    root = logging.getLogger()
    f = TraceLogFilter()
    for h in root.handlers:
        h.addFilter(f)
